metadata-ingestion/src/datahub/ingestion/source/state/redundant_run_skip_handler.py

```diff
@@ -104,46 +99,178 @@ def create_checkpoint(self) -> Optional[Checkpoint[BaseUsageCheckpointState]]:
+            self.job_id, BaseTimeWindowCheckpointState
         )
+
         if last_checkpoint and last_checkpoint.state:
-            state = cast(BaseUsageCheckpointState, last_checkpoint.state)
-            last_successful_pipeline_run_end_time_millis = state.end_timestamp_millis
+            last_run_start_time: int = last_checkpoint.state.begin_timestamp_millis
+            last_run_end_time: int = last_checkpoint.state.end_timestamp_millis
+
+            logger.debug(
+                f"{self.job_id} : Last run start, end times:"
+                f"({get_datetime_from_ts_millis_in_utc(last_run_start_time)}, {get_datetime_from_ts_millis_in_utc(last_run_end_time)})"
+            )
 
+            # If current run's time window is subset of last run's time window, then skip.
+            # Else there is at least some part in current time window that was not covered in past run's time window
+            if (
+                last_run_start_time <= cur_start_time_millis <= last_run_end_time
+                and last_run_start_time <= cur_end_time_millis <= last_run_end_time
+            ):
+                skip = True
+
+        return skip
+
+    def suggest_run_time_window(
+        self,
+        cur_start_time: datetime,
+        cur_end_time: datetime,
+        allow_reduce: int = True,
+        allow_expand: int = False,
+    ) -> Tuple[datetime, datetime]:
+        cur_start_time_millis = datetime_to_ts_millis(cur_start_time)
+        cur_end_time_millis = datetime_to_ts_millis(cur_end_time)
+        last_checkpoint = self.state_provider.get_last_checkpoint(
+            self.job_id, BaseTimeWindowCheckpointState
+        )
         if (
-            last_successful_pipeline_run_end_time_millis is not None
-            and cur_start_time_millis <= last_successful_pipeline_run_end_time_millis
-        ):
-            warn_msg = (
-                f"Skippig this run, since the last run's bucket duration end: "
-                f"{get_datetime_from_ts_millis_in_utc(last_successful_pipeline_run_end_time_millis)}"
-                f" is later than the current start_time: {get_datetime_from_ts_millis_in_utc(cur_start_time_millis)}"
+            last_checkpoint is None or last_checkpoint.state is None
+        ) or self.should_skip_this_run(cur_start_time_millis, cur_end_time_millis):
+            return cur_start_time, cur_end_time
+
+        suggested_start_time_millis: int = cur_start_time_millis
+        suggested_end_time_millis: int = cur_end_time_millis
+
+        last_run_start_time, last_run_end_time = self.get_last_run_time_window(
+            last_checkpoint
+        )
+
+        # In case of usage, it is assumed that start_time_millis (cur_start_time_millis as well as last_run_start_time) always conincides
+        # with floored bucket start time. This should be taken care of outside this scope.
+
+        if cur_start_time_millis >= last_run_start_time:
+            if cur_start_time_millis > last_run_end_time:
```

Comment:
I think this logic could be slightly easier to follow if we were more declarative on the cases:

```
if cur_start_time > last_run_end_time:
    if allow_expand: ...
    else: ...
elif (allow_reduce and last_run_start_time <= cur_start_time < last_run_end_time <= cur_end_time):
    suggested_start_time_millis = last_run_end_time
elif (allow_reduce and cur_start_time <= last_run_start_time < cur_end_time <= last_run_end_time):
    suggested_end_time_millis = last_run_start_time
```
What do you think?

Reply:
I am not sure. There are total 6 cases of time window here, and all these comparisons (in fact multiple interval comparisons on single line) look pretty confusing to me to catch up with. 

I think, the ideal approach to make it simpler to follow would be to define an object representation of time window and define its own methods such as 
1. whether current window is subset of another window.
2. whether current window overlaps with another window
3. whether current window comes before or after another window

and then declare logic using these methods. Thoughts ?